pkg/scheduler/util/assumecache/assume_cache.go

```diff
@@ -221,13 +249,26 @@ func (c *AssumeCache) add(obj interface{}) {
 			c.logger.V(10).Info("Skip adding object to assume cache because version is not newer than storedVersion", "description", c.description, "cacheKey", name, "newVersion", newVersion, "storedVersion", storedVersion)
 			return
 		}
+		oldObj = objInfo.latestObj
 	}
 
 	objInfo := &objInfo{name: name, latestObj: obj, apiObj: obj}
 	if err = c.store.Update(objInfo); err != nil {
 		c.logger.Info("Error occurred while updating stored object", "err", err)
 	} else {
 		c.logger.V(10).Info("Adding object to assume cache", "description", c.description, "cacheKey", name, "assumeCache", obj)
+		for _, handler := range c.eventHandlers {
```

Comment:
Seems we're maintaining two level event handlers here? I didn't quite get the idea why we need this? I read the comment of eventQueue, but still can't quite understand...

And do we have several handlers, what I learned from the PR is we only build the handler for `ResourceClaim`.

```
	for gvk, at := range gvkMap {		
                 case framework.ResourceClaim:
			if utilfeature.DefaultFeatureGate.Enabled(features.DynamicResourceAllocation) {
				// No need to wait for this cache to be
				// populated. If a claim is not yet in the
				// cache, scheduling will get retried once it
				// is.
				resourceClaimCache.AddEventHandler(buildEvtResHandler(at, framework.ResourceClaim, "ResourceClaim"))
			}
        }

Reply:
The chain of events is this now: claim informer -> assume cache -> plugin.

This ensures that local changes to the assume cache are properly reported to the plugin. This wasn't the case before, leading to the race explained in https://github.com/kubernetes/kubernetes/issues/123698

Reply:
Thanks for the explanation, I got the idea here, seems we have a MPSC architecture, we have producers like informer, local update, single consumer as the event handler. 

Based on that, maybe we'd better not consume the events immediately after producing them, but consuming them in another goroutine independently, that's say we'll move something like `emitEvents` to another place, decoupled with producer, which makes the code more readable and maintainable IMHO.

Moreover, we can hide the producer-consumer logic at cache layer, but locate them at the storage layer, take the update for example (this is just a demo) :

```
c.updateStore(objInfo) // rather than call c.store.Update() and push the events together.

func (c *AssumeCache) updateStore(oldObj, newObj) {
    c.store.Update()
    c.eventQueue.Push(oldObj, newObj)
}
```

My two cents.

Reply:
What you are proposing sounds *very* scary to me in terms of correctness and race conditions.

Can we keep the current architecture for event processing and just slightly extend it so that the source of events can be informers as well as local assume cache updates?

Reply:
I mean something like this https://github.com/kubernetes/kubernetes/commit/c7da77ca6350371a3e6f9977c420740438b30aa2, decouple the consumer logic.

Reply:
Can you explain why that is useful? What is the advantage of decoupling?

Event handlers are meant to complete quickly. That is already true for informer events. So making the code more complex to support long-running event handlers doesn't seem warranted to me.

Reply:
So the original idea is for simplify, the original code is a bit hard to understand, at least to me, so I don't think we're making it complex.

And I don't think it's slower, we'll consume the events continuously and we'll sent less events, on the contrary, I think we make less races, like 1) before when holding the lock, we'll iterate the handlers and send several related events, but now we only send one event, and 2) we'll call `emitEvents` at each OP, however, this will drain the whole events, which means we'll acquire unnecessary lock with several calls to `emitEvents` the same time.

BTW, for each OP, we'll send them to a queue and consume them from the queue in one function, what's your take here vs `invoking handlers directly`? Are you afraid the handlers will take a long time?

Would like to cc @alculquicondor @jsafrane for advices as well.

Reply:
Sorry, I am not following. The assume cache should emit the same number of events regardless in which goroutine the event handlers are called, so "now we only send one event" doesn't make sense to me.

What does "OP" stand for?

Reply:
What I mean is before, we loop the handlers to sent separated events like this during the lock time:
```
			for _, handler := range c.eventHandlers {
				handler := handler
				oldObj, obj := objInfo.latestObj, objInfo.apiObj
				c.eventQueue.Push(func() {
					handler.OnUpdate(oldObj, obj)
				})
			}
```

Now we do like this:
```
*Producer*: 
c.eventQueue.Push(event{OldObj: oldObj, Obj: obj})

*Consumer*:
for _, handler := range c.eventHandlers {
	handler := handler
	switch action {
	case "add":
		handler.OnAdd(event.Obj, false)
	case "update":
		handler.OnUpdate(event.OldObj, event.Obj)
	case "delete":
		handler.OnDelete(event.OldObj)
	}
}
``` 
So we're still doing the same thing, no handler will miss the event, but with different patterns.

> What does "OP" stand for?

Referring to the functions like `AddEventHandler`, `Restore`, `Assume`, `add/update/delete`.

Reply:
About "goroutines are more complex".

https://github.com/kubernetes/kubernetes/commit/c7da77ca6350371a3e6f9977c420740438b30aa2 currently has no "WaitForShutdown"-style API call, so unit tests cannot be sure that the goroutine has stopped. Lots of components in Kubernetes have that problem, but that doesn't mean that it's okay. When I write new code that spawns goroutines, I am making sure that cleanup and termination is handled. Can be added, of course.

The other drawback is that, as you noticed, unit tests are now no longer deterministic (`FIXME: Need to wait until consuming the events, it's a decoupled architecture.`). I'm curious how you want to fix that.

Reply:
I checked the code @pohly wrote, and while it's harder to follow the flow (informer -> assume cache -> plugin), I think it's usable.
I personally think c7da77c is adding only complications, without helping with code readability.